07.ls_object/file_display.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+require 'optparse'
+
+class FileDisplay
+  NUMBER_OF_COLUMNS = 3
+
+  def initialize
+    @options = ARGV.getopts('alr').transform_keys(&:to_sym)
+    @file_stats = fetch_file_stats
+  end
+
+  def select_format
+    @options[:l] ? long_format : short_format
+  end
+
+  private
+
+  def fetch_file_stats
+    flags = @options[:a] ? File::FNM_DOTMATCH : 0
+    file_names = Dir.glob('*', flags)
+    sorted_file_names = @options[:r] ? file_names.reverse : file_names
+    sorted_file_names.map { |file_name| FileStat.new(file_name) }
+  end
+
+  def long_format
+    blocks_total = @file_stats.map(&:blocks).sum
+    puts "total #{blocks_total}"
+
+    @file_stats.each { |file_stat| puts format_file_stat(file_stat) }
+  end
+
+  def format_file_stat(file_stat)
+    stat_map = build_stat_map(file_stat)
+    max_size_map = build_max_size_map(@file_stats)
+
+    format([
+      '%<file_mode>s',
+      "%<nlink>#{max_size_map[:nlink] + 1}s",
+      "%<user_name>-#{max_size_map[:user_name] + 1}s",
+      "%<group_name>-#{max_size_map[:group_name] + 1}s",
+      "%<size>#{max_size_map[:size]}s",
+      '%<mtime>2s',
+      "%<name>s\n"
+    ].join(' '), stat_map)
```

Comment:
一気にやらずに処理を分けましょう。
```suggestion
    hoge = [
      '%<file_mode>s',
      "%<nlink>#{max_size_map[:nlink] + 1}s",
      "%<user_name>-#{max_size_map[:user_name] + 1}s",
      "%<group_name>-#{max_size_map[:group_name] + 1}s",
      "%<size>#{max_size_map[:size]}s",
      '%<mtime>2s',
      "%<name>s\n"
    ].join(' ')
    format(hoge, stat_map)
```

また `build_stat_map` の定義を見に行かないと、`stat_map`が本当に`<file_mode>`や`<nlink>`とマッチしているかわからないので、こういうケースはあえてメソッドを分けない方がいいと思います。


```suggestion
    stat_map = {
      file_mode: file_stat.file_mode,
      nlink: file_stat.nlink,
      user_name: file_stat.user.name,
      group_name: file_stat.group.name,
      size: file_stat.size,
      mtime: file_stat.mtime.strftime('%_m %_d %H:%M'),
      name: file_stat.name
    }
    hoge = [
      '%<file_mode>s',
      "%<nlink>#{max_size_map[:nlink] + 1}s",
      "%<user_name>-#{max_size_map[:user_name] + 1}s",
      "%<group_name>-#{max_size_map[:group_name] + 1}s",
      "%<size>#{max_size_map[:size]}s",
      '%<mtime>2s',
      "%<name>s\n"
    ].join(' ')
    format(hoge, stat_map)
```

Reply:
> 一気にやらずに処理を分けましょう。

承知しました、一気にするよりも処理を分けたほうが見通しがよくなり、コードを理解しやすくなりますね。
こちら修正しました。

> また build_stat_map の定義を見に行かないと、stat_mapが本当に<file_mode>や<nlink>とマッチしているかわからないので、こういうケースはあえてメソッドを分けない方がいいと思います。

承知しました。
おっしゃる通り、定義したものとその定義したものを照らし合わせる必要がある場合は、同じメソッドにまとめたほうがよいですね。
こちら修正しました。